### PR TITLE
Replaced base64_client_credentials with client_id and client_secret

### LINKED ISSUE #5: Support client ID and secret in config
The config should allow the client ID and secret to be configured. The configuration of the basic HTTP header should be removed and this new config should be used instead.

### PR DESCRIPTION
**Pull request by anestos**
Fixes #5

### CODE REVIEW

File: ngx_http_access_token_to_jwt.c [hunk 1]
```diff
@@ -127,6 +137,20 @@ static ngx_int_t ngx_http_access_token_to_jwt_handler(ngx_http_request_t *reques
...
+    if (module_location_config->client_id.len == 0)
+    {
+        ngx_log_error(NGX_LOG_WARN, request->connection->log, 0,
+                           "Module not configured properly: missing client id");
+        return NGX_DECLINED;
+    }
+
     ngx_str_t encoded_client_credentials = module_location_config->base64encoded_client_credentials;
 
     if (encoded_client_credentials.len == 0)
```
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: This is probably redundant

Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: Ya, that's removed at https://github.com/curityio/nginx_introspect/pull/10/files#diff-015a4cf98d13f6f15adc8c7c8584f672L70. Lines 154 to 164 should be removed.

File: ngx_http_access_token_to_jwt.c [hunk 2]
```diff
@@ -414,8 +436,41 @@ static char *ngx_http_access_token_to_jwt_merge_loc_conf(ngx_conf_t *config, voi
...
 
+
+    ngx_str_t  *concat_credentials =  ngx_pcalloc(config->pool, sizeof(ngx_str_t));
+
+    if (concat_credentials == NULL)
+    {
+        return NGX_CONF_ERROR;
+    }
+
+    int concat_credentials_size = conf->client_id.len + conf->client_secret.len + 1; // client_id:client_secret
```
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: you don't leave room here for the nul byte. Do you need this? Usually you do, but NGIX does strings a little differently, I've seen.

```
client_id:client_secret0
0        1         2
123456789012345678901234
```

Using `client_id` and `client_secret` as the client ID and secret, `concat_credentials_size` will be 23 bytes long. To terminate it with a nul byte, you'll overrun your buffer.

File: ngx_http_access_token_to_jwt.c [hunk 3]
```diff
@@ -414,8 +436,41 @@ static char *ngx_http_access_token_to_jwt_merge_loc_conf(ngx_conf_t *config, voi
...
 
     ngx_conf_merge_value(conf->enable, prev->enable, 0);
-    ngx_conf_merge_str_value(conf->base64encoded_client_credentials, prev->base64encoded_client_credentials, "");
+
+    ngx_conf_merge_str_value(conf->client_id, prev->client_id, "");
+    ngx_conf_merge_str_value(conf->client_secret, prev->client_secret, "");
     ngx_conf_merge_str_value(conf->introspection_endpoint, prev->introspection_endpoint, "");
 
+
+    ngx_str_t  *concat_credentials =  ngx_pcalloc(config->pool, sizeof(ngx_str_t));
```
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: the old name of `base64_encoded_client_credentials` would work here. If you accept this suggested name change, I'd also suggest changing `concat_credentials_size` to `base64_encoded_client_credentials_size`.

File: ngx_http_access_token_to_jwt.c [hunk 4]
```diff
@@ -414,8 +436,41 @@ static char *ngx_http_access_token_to_jwt_merge_loc_conf(ngx_conf_t *config, voi
...
+
+    if (concat_credentials_data == NULL)
+    {
+        return NGX_CONF_ERROR;
+    }
+
+    concat_credentials->data = concat_credentials_data;
+    concat_credentials->len = concat_credentials_size;
+
+    ngx_snprintf(concat_credentials_data, concat_credentials_size, "%s:%s", conf->client_id.data, conf->client_secret.data);
```
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: from the `vsnprintf` man page:

> The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string (the size'th character then gets the terminating `\0');

So, ya, you're not gonna overrun your buffer, but it looks like you'll chop off the last character in the password.

Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: You can use `%V` here, so:

```c
 ngx_snprintf(..., "%V:%V", conf->client_id, conf->client_secret);
```

File: ngx_http_access_token_to_jwt.c [hunk 5]
```diff
@@ -414,8 +436,41 @@ static char *ngx_http_access_token_to_jwt_merge_loc_conf(ngx_conf_t *config, voi
...
+
+    conf->base64encoded_client_credentials.data = ngx_pcalloc(config->pool, ngx_base64_encoded_length(concat_credentials_size));
+
+    if (conf->base64encoded_client_credentials.data == NULL)
+    {
+        return NGX_CONF_ERROR;
+    }
+
+    ngx_encode_base64(&conf->base64encoded_client_credentials, concat_credentials);
```
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: you can free `concat_credentials` at this point using `ngx_pfree(config->pool, concat_credentials)`.

File: ngx_http_access_token_to_jwt.c (hunk 1 above)
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: Ah, ya, I see. It's still used but assigned in the merge function below. This is actually a programming error if the value isn't there.

File: ngx_http_access_token_to_jwt.c (hunk 4 above)
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: nginx doesn't start if i do this change (segmentation fault).

File: ngx_http_access_token_to_jwt.c (hunk 5 above)
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: added

File: ngx_http_access_token_to_jwt.c (hunk 2 above)
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: I don't see any problem with that. The header that is sent to Curity is correct for every client_id and client_secret combination i checked

File: ngx_http_access_token_to_jwt.c (hunk 3 above)
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: I aggree, the name is not great, but renaming `concat_credentials` to `base64_encoded_client_credentials` would be misleading . It contains the plain `client_id:client_secret` after `ngx_snprintf(concat_credentials_data, concat_credentials_size, "%s:%s", conf->client_id.data, conf->client_secret.data);`

The encoded version of that string is saved in the `base64encoded_client_credentials` variable of the configuration directly.

Any other suggestions on the rename?

File: ngx_http_access_token_to_jwt.c (hunk 1 above)
Comment (anestos) [inline ngx_http_access_token_to_jwt.c]: Even for empty configuration on client_id and client_secret this is not empty ( has the value `Og==` which is the base64 encoded `:`). I ll remove this check

File: ngx_http_access_token_to_jwt.c (hunk 3 above)
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: ah, thanks for clarifying. What about `unencoded_client_credentials`?

File: ngx_http_access_token_to_jwt.c (hunk 4 above)
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: ah, ` %V — ngx_str_t * ` It's a pointer, so you'd have to change to this:
```c
ngx_snprintf(..., "%V:%V", &conf->client_id, &conf->client_secret);
```
Sorry. learning here.

File: ngx_http_access_token_to_jwt.c [hunk 6]
```diff
@@ -153,15 +153,6 @@ static ngx_int_t ngx_http_access_token_to_jwt_handler(ngx_http_request_t *reques
 
     ngx_str_t encoded_client_credentials = module_location_config->base64encoded_client_credentials;
 
-    if (encoded_client_credentials.len == 0)
```
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: maybe add `assert(encoded_client_credentials.len > 0)`

File: ngx_http_access_token_to_jwt.c (hunk 2 above)
Comment (travisspencer) [inline ngx_http_access_token_to_jwt.c]: I tested it. You were getting lucky is all. You can see this if you add `concat_credentials[concat_credentials_size] = 1;` right after you alloc the memory for `concat_credentials`. In this case, you'll always have an ASCII Start of Header (SOH) char at the end of your password. This is fixed in 91c44ac where I added a macro, `basic_credential_length` to calculate the length of the header like this:

```c
#define basic_credential_length(id, secret) id + (sizeof(":") - 1) + secret + (sizeof("\0") - 1)
```

With this fixed and the other stuff you did, this branch is ready to be merged once the conflicts are resolved. I'll leave that to you.

**Pull request merged**